// points/cjs/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.numberAtInterval = exports.isShapeArray = exports.linearLength = exports.getShapeArray = exports.applyFuncToShapes = exports.angleFromSides = void 0;

var angleFromSides = function angleFromSides(a, b, c) {
  var r = Math.acos((Math.pow(a, 2) + Math.pow(b, 2) - Math.pow(c, 2)) / (2 * a * b));
  return r * (180 / Math.PI);
};

exports.angleFromSides = angleFromSides;

var applyFuncToShapes = function applyFuncToShapes(f, s) {
  for (var _len = arguments.length, args = new Array(_len > 2 ? _len - 2 : 0), _key = 2; _key < _len; _key++) {
    args[_key - 2] = arguments[_key];
  }

  if (isShapeArray(s)) {
    return s.map(function (shape) {
      return f.apply(void 0, [shape].concat(args));
    });
  }

  return f.apply(void 0, [s].concat(args));
};

exports.applyFuncToShapes = applyFuncToShapes;

var getShapeArray = function getShapeArray(s) {
  return isShapeArray(s) ? s : [s];
};

exports.getShapeArray = getShapeArray;

var linearLength = function linearLength(x1, y1, x2, y2) {
  return Math.sqrt(Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2));
};

exports.linearLength = linearLength;

var isShapeArray = function isShapeArray(s) {
  return Array.isArray(s[0]);
};

exports.isShapeArray = isShapeArray;

var numberAtInterval = function numberAtInterval(a, b, interval) {
  var c = a === b ? 0 : Math.abs(b - a);
  return c === 0 ? a : a < b ? a + c * interval : a - c * interval;
};

exports.numberAtInterval = numberAtInterval;
