// points/cjs/reverse.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _cubify = _interopRequireDefault(require("./cubify"));

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var reversePoints = function reversePoints(shape) {
  var m;
  var c;
  return shape.reverse().map(function (_ref, i) {
    var x = _ref.x,
        y = _ref.y,
        moveTo = _ref.moveTo,
        curve = _ref.curve;
    var point = {
      x: x,
      y: y
    };

    if (c) {
      var _c = c,
          x2 = _c.x1,
          y2 = _c.y1,
          x1 = _c.x2,
          y1 = _c.y2;
      point.curve = {
        type: 'cubic',
        x1: x1,
        y1: y1,
        x2: x2,
        y2: y2
      };
    }

    if (i === 0 || m) {
      point.moveTo = true;
    }

    m = moveTo;
    c = curve || null;
    return point;
  });
};

var reverse = function reverse(s) {
  return (0, _helpers.applyFuncToShapes)(reversePoints, (0, _cubify["default"])(s));
};

var _default = reverse;
exports["default"] = _default;
